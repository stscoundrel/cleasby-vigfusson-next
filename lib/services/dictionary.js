const { getDictionary } = require('cleasby-vigfusson-dictionary');
const slugify = require('slugify');

const addSlugs = (words) => {
  const existingSlugs = {};

  const formattedWords = words.map((word) => {
    let slug = slugify(word.word).toLowerCase();

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1;
      slug = `${slug}-${existingSlugs[slug]}`;
    } else {
      existingSlugs[slug] = 1;
    }

    return {
      ...word,
      slug,
    };
  });

  return formattedWords;
};

export const getAllWords = () => {
  const words = getDictionary();

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words);

  return formattedWords;
};

export default {
  getAllWords,
};
