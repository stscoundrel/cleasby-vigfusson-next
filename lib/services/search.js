export const searchDictionary = (search, dictionary) => {
  const results = dictionary.filter((entry) => {
    let matchesSearch = false

    if (entry.word.includes(search)) {
      matchesSearch = true
    }

    if (entry.slug.includes(search)) {
      matchesSearch = true
    }

    entry.definitions.forEach((definition) => {
      if (definition.includes(search)) {
        matchesSearch = true
      }
    })

    return matchesSearch
  })

  return results
}

export default {
  searchDictionary,
}
