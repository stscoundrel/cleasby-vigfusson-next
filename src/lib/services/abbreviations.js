import { findAbbreviations, findWorksAndAuthors } from 'cleasby-vigfusson-abbreviations'
import { abbreviate } from 'abbreviatrix'

/**
 * Combine abbreviations from definitions array.
 * Accepts finder method to use.
 */
const combineAbbreviations = ({ definitions }, find) => {
  const combinedAbbreviations = []
  const abbreviationSet = new Set()

  definitions.forEach((definition) => {
    const abbreviations = find(definition)
    abbreviations.forEach((explanation, abbreviation) => {
      if (!abbreviationSet.has(abbreviation)) {
        abbreviationSet.add(abbreviation)
        combinedAbbreviations.push({ abbreviation, explanation })
      }
    })
  })

  return combinedAbbreviations;
}

export const getAbbreviations = (entry) => ({
  common: combineAbbreviations(entry, findAbbreviations),
  works: combineAbbreviations(entry, findWorksAndAuthors),
})

/**
 * Add abbr tags to content with explanations.
 */
export const addAbbreviationsToContent = (content, abbreviations) => {
  const combinedAbbrs = [...abbreviations.common, ...abbreviations.works]
  let result = content

  combinedAbbrs.forEach(({ abbreviation, explanation }) => {
    result = abbreviate(abbreviation, explanation, result)
  })

  return result
}

export default {
  getAbbreviations,
  addAbbreviationsToContent,
}
