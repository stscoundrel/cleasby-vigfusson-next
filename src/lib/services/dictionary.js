const { getDictionary } = require('cleasby-vigfusson-dictionary')
const { VALID_AS_FIRST } = require('old-norse-alphabet')
const { slugifyWord, slugifyLetter } = require('../utils/slugs')

let cachedDictionary = null

const addSlugs = (words) => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (slug.charAt(0) === '-') {
      slug = slug.substring(1)
    }

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = () => {
  if (cachedDictionary) return cachedDictionary

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  cachedDictionary = formattedWords

  return formattedWords
}

export const getByLetter = (letter) => {
  const words = getAllWords()
  const byLetter = words.filter((entry) => (
    entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))

  return byLetter
}

export const getWord = (word) => getAllWords().filter((entry) => entry.slug === word)[0]

export const getAlphabet = () => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getAlphabet,
}
