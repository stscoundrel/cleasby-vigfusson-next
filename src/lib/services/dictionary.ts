import { getDictionary, DictionaryEntry as RawDictionaryEntry } from 'cleasby-vigfusson-dictionary'
import { VALID_AS_FIRST } from 'old-norse-alphabet'
import { oldNorseSort } from 'old-norse-alphabet-sort'
import { slugifyWord, slugifyLetter } from '../utils/slugs'

export interface DictionaryEntry extends RawDictionaryEntry {
  slug: string,
}

export interface DictionaryEntryDTO {
  word: string,
  slug: string
}

export interface AlphabetLetter {
  letter: string,
  slug: string
}

let cachedDictionary: DictionaryEntry[] | null = null

const addSlugs = (words: RawDictionaryEntry[]): DictionaryEntry[] => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (slug.charAt(0) === '-') {
      slug = slug.substring(1)
    }

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = (): DictionaryEntry[] => {
  if (cachedDictionary) return cachedDictionary

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  /**
   * Ensure headwords are lowercased.
   * Original source has some in all caps.
   * That is true to the layout of the book,
   * but makes little sense in website.
   */
  const dictionary = formattedWords.map((entry) => ({ ...entry, word: entry.word.toLowerCase() }))

  cachedDictionary = dictionary

  return dictionary
}

export const getByLetter = (letter: string): DictionaryEntryDTO[] => {
  const words = getAllWords()
  const byLetter = words
    .filter((entry) => (
      entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))
    .map((entry) => {
      // Simpler DTO dictionary entry.
      const { word, slug } = entry
      return { word, slug }
    })
    .sort((a, b) => oldNorseSort(a.word, b.word))

  return byLetter
}

export const getWord = (slug: string): DictionaryEntry => (
  getAllWords().filter((entry) => entry.slug === slug)[0]
)

export const getRandomEntries = (): DictionaryEntry[] => (
  // Return entries fit to be randomized "teasers"
  // Therefore, content should be short, but not too short.
  getAllWords()
    .sort(() => Math.random() - 0.5)
    .filter((entry) => entry.definitions[0].length < 50 && entry.definitions[0].length > 15)
    .slice(0, 36)
    .sort((a, b) => oldNorseSort(a.word, b.word))
)

export const getSimilarWords = (entry: DictionaryEntry) : DictionaryEntry[] => getAllWords()
  .filter((dEntry) => dEntry.word.toLocaleLowerCase() === entry.word.toLocaleLowerCase()
  && dEntry.slug !== entry.slug)

export const getAlphabet = (): AlphabetLetter[] => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getRandomEntries,
  getAlphabet,
}
