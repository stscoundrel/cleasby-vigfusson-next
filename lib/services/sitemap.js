import { getAllWords, getAlphabet } from 'lib/services/dictionary'
import slugify from 'slugify'

const formatWords = (words) => {
  const siteUrl = process.env.NEXT_PUBLIC_SITE_URL

  return words.map((word) => ({
    url: `${siteUrl}/word/${word.slug}`,
    changefreq: 'monthly',
    priority: 0.5,
  }))
}

const formatLetters = (letters) => {
  const siteUrl = process.env.NEXT_PUBLIC_SITE_URL

  return letters.map((letter) => ({
    url: `${siteUrl}/letter/${slugify(letter)}`,
    changefreq: 'monthly',
    priority: 0.5,
  }))
}

export const getSitemap = () => {

}

export const getContent = () => {
  const words = formatWords(getAllWords())
  const letters = formatLetters(getAlphabet())

  return [...words, ...letters]
}

export default {
  getSitemap,
  getContent,
}
