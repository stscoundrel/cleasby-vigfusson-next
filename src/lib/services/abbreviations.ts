import abbreviationsService from 'cleasby-vigfusson-abbreviations'
import { abbreviate } from 'abbreviatrix'
import { DictionaryEntry } from 'lib/services/dictionary'

const { findAbbreviations, findWorksAndAuthors, getWorksAndAuthorsMapping } = abbreviationsService

export interface Abbreviation{
  abbreviation: string,
  explanation: string
}

export interface CombinedAbbreviations {
  common: Abbreviation[],
  works: Abbreviation[],
}

/**
 * Combine abbreviations from definitions array.
 * Accepts finder method to use.
 */
const combineAbbreviations = (
  { definitions }: DictionaryEntry,
  abbreviationFinder: (s: string) => Map<string, string>,
): Abbreviation[] => {
  const combinedAbbreviations: Abbreviation[] = []
  const abbreviationSet = new Set()

  definitions.forEach((definition) => {
    const abbreviations = abbreviationFinder(definition)
    abbreviations.forEach((explanation, abbreviation) => {
      if (!abbreviationSet.has(abbreviation)) {
        abbreviationSet.add(abbreviation)
        combinedAbbreviations.push({ abbreviation, explanation })
      }
    })
  })

  return combinedAbbreviations;
}

export const getAbbreviations = (entry: DictionaryEntry): CombinedAbbreviations => ({
  common: combineAbbreviations(entry, findAbbreviations),
  works: combineAbbreviations(entry, findWorksAndAuthors),
})

export const getAllSorces = (): Abbreviation[] => {
  const abbrs = getWorksAndAuthorsMapping()
  return Array
    .from(abbrs, ([abbreviation, explanation]) => ({ explanation, abbreviation }))
    .sort((a, b) => a.abbreviation.localeCompare(b.abbreviation))
}

/**
 * Add abbr tags to content with explanations.
 */
export const addAbbreviationsToContent = (
  content: string,
  abbreviations: CombinedAbbreviations,
): string => {
  const combinedAbbrs = [...abbreviations.common, ...abbreviations.works]
  let result = content

  combinedAbbrs.forEach(({ abbreviation, explanation }) => {
    result = abbreviate(abbreviation, explanation, result)
  })

  return result
}

export default {
  getAbbreviations,
  addAbbreviationsToContent,
}
