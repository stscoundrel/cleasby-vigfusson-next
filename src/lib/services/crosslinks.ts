import { getCrosslinks as getAllCrossLinks, Crosslink, getOldNorseCrosslinks } from 'scandinavian-dictionary-crosslinker'
import { DictionaryEntry } from 'lib/services/dictionary'

export type { Crosslink } from 'scandinavian-dictionary-crosslinker'

let crossLinkCache: Record<string, Crosslink[]> | null = null

const populateCrossLinks = () => {
  if (!crossLinkCache) {
    crossLinkCache = getAllCrossLinks()
  }

  return crossLinkCache
}

export const getCrossLinks = (entry: DictionaryEntry): Crosslink[] => {
  const crosslinks = populateCrossLinks()
  if (Object.prototype.hasOwnProperty.call(crosslinks, entry.slug)) {
    return getOldNorseCrosslinks(entry.slug)
  }

  return []
}
