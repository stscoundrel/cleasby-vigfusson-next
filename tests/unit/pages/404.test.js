import ReactDOM from 'react-dom/client'
import Page404, { getStaticProps } from 'pages/404'
import renderer from 'react-test-renderer'
import { getAlphabet } from 'lib/services/dictionary'

describe('404 page page', () => {
  test('Does not crash', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(<Page404 letters={getAlphabet()} />)
  })

  test('Matches snapshot', () => {
    const tree = renderer.create(<Page404 letters={getAlphabet()} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('404 page: data fetching', () => {
  test('getStaticProps works', async () => {
    const expected = {
      props: {
        letters: getAlphabet(),
      },
    }

    const result = await getStaticProps()

    expect(result).toEqual(expected)
  })
})
