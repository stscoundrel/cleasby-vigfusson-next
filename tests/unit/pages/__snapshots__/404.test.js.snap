// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`404 page page Matches snapshot 1`] = `
[
  <header>
    <div
      className="topBarSpacer"
    />
    <nav
      className="section"
    >
      <div
        className="topbar container"
      >
        <img
          alt="To home"
          height="30"
          loading="lazy"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
          src="/favicon-48x48.png"
          width="30"
        />
        <div
          aria-label="Open menu"
          className="hamburger"
          onClick={[Function]}
          role="button"
        >
          <span
            className="part"
          />
          <span
            className="part"
          />
          <span
            className="part"
          />
        </div>
      </div>
      <div
        className="content null container"
      >
        <ul
          className="list"
        >
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/a"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              a
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/a2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              á
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/b"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              b
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/d"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              d
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/e"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              e
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/e2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              é
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/f"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              f
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/g"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              g
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/h"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              h
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/i"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              i
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/i2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              í
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/j"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              j
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/k"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              k
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/l"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              l
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/m"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              m
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/n"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              n
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/o"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              o
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/o2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ó
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/p"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              p
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/r"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              r
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/s"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              s
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/t"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              t
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/u"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              u
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/u2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ú
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/v"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              v
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/y"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              y
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/y2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ý
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/th"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              þ
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/ae"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              æ
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/oe3"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              œ
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/oe"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ö
            </a>
          </li>
        </ul>
        <form
          className="form"
          onSubmit={[Function]}
        >
          <input
            aria-label="Search"
            className="input"
            name="search"
            onChange={[Function]}
            type="search"
            value=""
          />
          <button
            className="button button--secondary button--no-margs"
            name="submit"
            type="submit"
          >
            Search
          </button>
        </form>
      </div>
    </nav>
    <nav
      className="section"
    >
      <div
        className="container"
      >
        <a
          className="link"
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Cleasby & Vigfusson Dictionary
        </a>
        <script
          dangerouslySetInnerHTML={
            {
              "__html": "{"@context":"https://schema.org/","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Cleasby & Vigfusson Dictionary","item":"undefined/"}]}",
            }
          }
          type="application/ld+json"
        />
      </div>
    </nav>
  </header>,
  <main
    className="container"
  >
    <section
      className="section"
    >
      <h1>
        Page not found
      </h1>
      <p>
        Huh, that's weird.
      </p>
    </section>
    <div
      aria-label="Back to top"
      className="backToTop"
      onClick={[Function]}
      role="button"
    >
      ↑
    </div>
  </main>,
  <footer
    className="section"
  >
    <div
      className="container"
    >
      <section
        className="section"
      >
        <h2>
          About
        </h2>
        <p>
          Based on Cleasby & Vigfusson Dictionary Old Norse dictionary.
        </p>
        <p>
          <em>
            Icelandic-English
          </em>
           dictionary was started by Richard Cleasby and finished by Gudbrand Vigfusson.
        </p>
        <p>
          It was published in 1874, which leads to there being many public domain versions of the book available.
        </p>
      </section>
      <section
        className="section"
      >
        <h3>
          Old Norse language
        </h3>
        <p>
          Old Norse was a North Germanic language that was spoken by inhabitants of Scandinavia and their overseas settlements from about the 7th to the 15th centuries.
        </p>
        <p>
          Also known as "the viking language", "Old Nordic", or "Old Scandinavian"
        </p>
      </section>
      <div
        className="navs"
      >
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Dictionary project
          </h4>
          <ul>
            <li>
              <a
                href="https://github.com/stscoundrel/cleasby-vigfusson-next"
                rel="noopener noreferrer"
                target="_blank"
              >
                Source code
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/cleasby-vigfusson-dictionary"
                rel="noopener noreferrer"
                target="_blank"
              >
                Data source
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/cleasby-vigfusson-abbreviations"
                rel="noopener noreferrer"
                target="_blank"
              >
                Abbreviations
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/younger-futhark"
                rel="noopener noreferrer"
                target="_blank"
              >
                Younger Futhark Runes
              </a>
            </li>
          </ul>
        </nav>
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Related dictionary projects
          </h4>
          <ul>
            <li>
              <a
                href="https://old-icelandic.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                A Concise Dictionary of Old Icelandic
              </a>
            </li>
            <li>
              <a
                href="https://old-swedish-dictionary.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                K.F Söderwall's Old Swedish Dictionary
              </a>
            </li>
            <li>
              <a
                href="https://old-norwegian-dictionary.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Dictionary of the Old Norwegian Language
              </a>
            </li>
            <li>
              <a
                href="https://old-danish-dictionary.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Otto Kalkar's Old Danish Dictionary
              </a>
            </li>
          </ul>
        </nav>
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Quick links
          </h4>
          <ul
            className="navColumns"
          >
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/a"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                a
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/a2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                á
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/b"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                b
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/d"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                d
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/e"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                e
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/e2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                é
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/f"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                f
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/g"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                g
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/h"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                h
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/i"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                i
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/i2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                í
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/j"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                j
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/k"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                k
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/l"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                l
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/m"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                m
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/n"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                n
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/o"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                o
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/o2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ó
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/p"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                p
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/r"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                r
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/s"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                s
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/t"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                t
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/u"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                u
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/u2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ú
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/v"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                v
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/y"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                y
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/y2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ý
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/th"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                þ
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/ae"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                æ
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/oe3"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                œ
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/oe"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ö
              </a>
            </li>
          </ul>
        </nav>
      </div>
      <small
        className="copyright"
      >
        Copyright © 2020 - 2020
        <br />
        <a
          href="https://github.com/stscoundrel"
          rel="noopener noreferrer"
          target="_blank"
        >
          Sampo Silvennoinen / StScoundrel
        </a>
      </small>
    </div>
  </footer>,
]
`;
